diff --git a/src/airconv/darwin/meson.build b/src/airconv/darwin/meson.build
index f3e39ec..b886b5f 100644
--- a/src/airconv/darwin/meson.build
+++ b/src/airconv/darwin/meson.build
@@ -19,6 +19,11 @@ if native_llvm_path.startswith('/usr/local/opt')
     llvm_ld_flags_darwin,
     '/usr/local/opt/zstd/lib/libzstd.a', '/usr/local/opt/llvm@15/lib/libunwind.a'
   ]
+elif native_llvm_path.startswith('/opt/homebrew/opt')
+  llvm_ld_flags_darwin = [
+    llvm_ld_flags_darwin,
+    '/opt/homebrew/opt/zstd/lib/libzstd.a', '/opt/homebrew/opt/llvm@15/lib/libunwind.a'
+  ]
 endif

 airconv_lib_darwin = static_library('airconv', airconv_src,
